File: packages/hydris/src/core/CoreServiceProvider.ts
import { DiContainer } from '../container/DiContainer';
import { ServiceProvider } from '../container/ServiceProvider';
import { CookieAuthority } from '../http/CookieAuthority';
import { Request } from '../http/Request';
import { Csrf } from '../middleware/Csrf';
import { MatchedRoute } from '../routing/MatchedRoute';

export class CoreServiceProvider extends ServiceProvider {
    private static readonly HTTP_SCOPE_HINT =
        'these tokens are only available inside an HTTP request scope';

    register(container: DiContainer): void {
        container.contextTokens(
            'http',
            [Request, MatchedRoute],
            CoreServiceProvider.HTTP_SCOPE_HINT,
        );
        container.scoped(
            CookieAuthority,
            (c) => new CookieAuthority(c.get(Request)),
        );
        container.transient(Csrf, (c) => new Csrf(c.get(CookieAuthority)));
    }

    static requestScope(container: DiContainer, req: Request): DiContainer {
        const scope = container.scope();
        scope.value(DiContainer, scope);
        scope.value(Request, req);
        if (req.route) scope.value(MatchedRoute, req.route);
        return scope;
    }
}

File: packages/hydris/Hydris/Core/CoreServiceProvider.cs
using System.IO;
using Hydris.Container;
using Hydris.Http;

namespace Hydris.Core;

public sealed class CoreServiceProvider : ServiceProvider {
    private const string HttpScopeHint = "these tokens are only available inside an HTTP request scope";

    public override void Register(DiContainer container) {
        container.ContextTokens("http", [typeof(Request)], HttpScopeHint);
        container.Singleton(_ => ElemixManifest.Load(
            Path.Combine(AppContext.BaseDirectory, "public", "_elemix", "manifest.json")));
        container.Scoped(c => new CookieAuthority(c.Get<Request>()));
    }

    public static DiContainer RequestScope(DiContainer container, Request request) {
        var scope = container.Scope();
        scope.Value(scope);
        scope.Value(request);
        return scope;
    }
}
